README
========
 Butter
========
Butter is a library to integrate some of Linux's low level features into python
eg signalfd, eventfd and timerfd. most of these functions are handy if you are 
looking into creating non-blocking servers, anything using select/poll and 
derivatives or writing server code

Whats Available
----------------
 * inotify (Complete support, includes asyncio support)
 * seccomp (Limited support)
 * fanotify (Limited support, includes asyncio support)
 * splice
 * tee
 * vmsplice
 * gethostname
 * sethostname
 * mount
 * umount
 * pivot_root
 * getpid (bypasses glibc caching)
 * getppid (bypasses glibc caching)
 * eventfd (includes asyncio support)
 * timerfd (includes asyncio support)
 * pthread_sigmask (Avalible in python3.x but backported for python2.7)
 * signalfd (includes asyncio support)

Whats Coming
-------------
Most of these exist in v0.2 as ctypes code. these are currently being rewritten
to use cffi for speed and compatibility with pypy

 * posix/linux aio (scatter/gather read writes with a completion based API 
   instead of a 'ready' based interface, eg similar to IOCP on windows)

Supported Python
-----------------
Butter currently supported the following python interpreters

* pypy (2.7 and 3.2 python implementations)
* cpython 3.4 (required for asyncio support)
* cpython 3.x
* cpython 2.7

Butter may work on older versions however it has not been tested on anything 
except the above interpreters and may break without warning

Design
-------
The 'fd' apis have been designed first as thin wrappers around the underlying
kernel syscalls. A File-like object that then uses these syscalls is then made 
available that obeys the 'event like' interface and is based on the interface
provided by File and friends. They are intended to be drop in replacements and 
be used with the select module

Event-like objects
+++++++++++++++++++
Event like objects are identical to filelike objects except they have the 
following extra methods to replace the disabled read/readline/write/writeline
functionality

 * read_event:  Return a single event
 * read_events: Return all cached events OR read all events from the kernel

AsyncIO
++++++++
Some syscalls (ones that deal with fd's) have been made to work with the 
asyncio framework that baecame avalible in python3.4. The asyncio enabled
object is named after the normal Event-like object equivlent with '_async'
after it, eg Inotify => Inotify_async. These asyncio Eventlike objects are
designed to act and behave like the asyncio.Queue object without the ability
to write events to the queue (ie they can only be read from and events are
injected from the fd as required)



CHANGELOG
butter (0.9-1) stable; urgency=low
  [ Da_Blitz ]
  * Add eventfd support
  * Add eventfd asyncio support
  * asyncio objects now have a close() method
  * eventfd now has a close() method
  * eventfd now has a fileno() method
  * Add timerfd support
  * Add timerfd asyncio support
  * Fixed issue with circular imports preventing python3.4 from working
  * Fixed issue with python2.7 returning floats where python3 returned ints
  * Added pthread_sigmask
  * Added Signalfd
  * Added Signalfd asyncio support
   
 -- Da_Blitz <code@pocketnix.org>  Tue, 17 May 2014 01:10:43 +1000

butter (0.8-1) stable; urgency=low
  [ Da_Blitz ]
  * Now works with python3.4 and higher
  * 'from butter import *' now imports the system module
  * Added trove classification
  * Added friendly properties to inotify event object
  * Added friendly properties to fanotify event object
  * FanotifyEvents now use less memory
  * AsyncIO support for inotify on supported platforms
  * AsyncIO support for fanotify on supported platforms
   
 -- Da_Blitz <code@pocketnix.org>  Tue, 17 May 2014 01:10:43 +1000

butter (0.7-1) stable; urgency=low
  [ Da_Blitz ]
  * Added system.py module
  * Added gethostname syscall
  * Added sethostname syscall
  * Added mount syscall
  * Added umount syscall
  * Added pivot_root syscall
  * Added getpid syscall
  * Added getppid syscall
  * Documented all new syscalls
   
 -- Da_Blitz <code@pocketnix.org>  Tue, 16 Mar 2014 22:04:43 +1000

butter (0.6-1) stable; urgency=low
  [ Da_Blitz ]
  * splice syscall documentation
  * Added tee() syscall
  * Added tee() example
  * Added vmsplice() syscall
  * Added vmsplice() example
  * Updated setup.py to newer auto detecting version
  * hide 'main' functions in splice module
 
 -- Da_Blitz <code@pocketnix.org>  Tue, 12 Mar 2014 23:26:43 +1000

butter (0.5-1) stable; urgency=low
  [ Da_Blitz ]
  * Added splice() syscall
 
 -- Da_Blitz <code@pocketnix.org>  Tue, 11 Mar 2014 02:11:43 +1000

butter (0.4-1) stable; urgency=low
  [ Da_Blitz ]
  * Refactor fanotify
  * Refactor inotify
  * Provide fanotify.str_to_events()
  * Provide inotify.str_to_events()
  * Add int to signal name mapping for inotify
 
 -- Da_Blitz <code@pocketnix.org>  Sun, 12 Dec 2013 11:10:43 +1000

butter (0.3-1) stable; urgency=low

  [ Da_Blitz ]
  * Support for inotify
  * Initial support for fanotify
  * Initial support for seccomp
  * Add function to peer inside kernel buffer and get amount of available bytes to read
  * removed unused old (non working) signalfd, eventfd, aio
  
 -- Da_Blitz <code@pocketnix.org>  Sun, 20 Nov 2013 22:31:43 +1000

butter (0.2-1) stable; urgency=low

  [ Da_Blitz ]
  * Initial support for signalfd
  * Initial support for eventfd
  * Initial support for aio
  
 -- Da_Blitz <code@pocketnix.org>  Sun, 20 Nov 2013 21:53:43 +1000



butter/errors.py
#!/usr/bin/env python
"""Basic libc error codes"""
import ctypes
import ctypes.util

libc = ctypes.CDLL(ctypes.util.find_library("c"))

class CError(Exception):
	"""Standard C Error
	"""
	def __init__(self, value, name, msg):
		self.value, self.name, self.msg = value, name, msg

	def __str__(self):
		return "{0} {1}: {2}".format(self.value, self.name, self.msg)

	def __repr__(self):
		return "<CError err_num={0}>".format(self.name)

def get_error(val):
	"""Returns a triplet that mapps to an error

	val can be an error name or number"""
	try:
		num = int(val)
		if num < 0:
			num = -num
		name, msg = ERRNO[num]				
	except ValueError:
		name = val
		try:
			num, msg = ERRNAME[name]
		except KeyError:
			# try again with E in front for lazy programmers
			num, msg = ERRNAME["E" + name]			

	return (num, name, msg)	

def check_error(val, exp_obj=CError):
	if val < 0:
		#errnum = ctypes.get_errno()
		errnum = ctypes.c_int.in_dll(libc, "errno").value
		num, name, msg = get_error(errnum)
		raise exp_obj(num, name, msg)


ERRNO = {
1:("EPERM", "Operation not permitted"),
2:("ENOENT", "No such file or directory"),
3:("ESRCH", "No such process"),
4:("EINTR", "Interrupted system call"),
5:("EIO", "I/O error"),
6:("ENXIO", "No such device or address"),
7:("E2BIG", "Argument list too long"),
8:("ENOEXEC", "Exec format error"),
9:("EBADF", "Bad file number"),
10:("ECHILD", "No child processes"),
11:("EAGAIN", "Try again/Operation would block"),
12:("ENOMEM", "Out of memory"),
13:("EACCES", "Permission denied"),
14:("EFAULT", "Bad address"),
15:("ENOTBLK", "Block device required"),
16:("EBUSY", "Device or resource busy"),
17:("EEXIST", "File exists"),
18:("EXDEV", "Cross-device link"),
19:("ENODEV", "No such device"),
20:("ENOTDIR", "Not a directory"),
21:("EISDIR", "Is a directory"),
22:("EINVAL", "Invalid argument"),
23:("ENFILE", "File table overflow"),
24:("EMFILE", "Too many open files"),
25:("ENOTTY", "Not a typewriter"),
26:("ETXTBSY", "Text file busy"),
27:("EFBIG", "File too large"),
28:("ENOSPC", "No space left on device"),
29:("ESPIPE", "Illegal seek"),
30:("EROFS", "Read-only file system"),
31:("EMLINK", "Too many links"),
32:("EPIPE", "Broken pipe"),
33:("EDOM", "Math argument out of domain of func"),
34:("ERANGE", "Math result not representable"),
35:("EDEADLK", "Resource deadlock would occur"),
36:("ENAMETOOLONG", "File name too long"),
37:("ENOLCK", "No record locks available"),
38:("ENOSYS", "Function not implemented"),
39:("ENOTEMPTY", "Directory not empty"),
40:("ELOOP", "Too many symbolic links encountered"),
#EWOULDBLOCK - EAGAIN	# Operation would block
42:("ENOMSG", "No message of desired type"),
43:("EIDRM", "Identifier removed"),
44:("ECHRNG", "Channel number out of range"),
45:("EL2NSYNC", "Level 2 not synchronized"),
46:("EL3HLT", "Level 3 halted"),
47:("EL3RST", "Level 3 reset"),
48:("ELNRNG", "Link number out of range"),
49:("EUNATCH", "Protocol driver not attached"),
50:("ENOCSI", "No CSI structure available"),
51:("EL2HLT", "Level 2 halted"),
52:("EBADE", "Invalid exchange"),
53:("EBADR", "Invalid request descriptor"),
54:("EXFULL", "Exchange full"),
55:("ENOANO", "No anode"),
56:("EBADRQC", "Invalid request code"),
57:("EBADSLT", "Invalid slot"),

#EDEADLOCK - EDEADLK

59:("EBFONT", "Bad font file format"),
60:("ENOSTR", "Device not a stream"),
61:("ENODATA", "No data available"),
62:("ETIME", "Timer expired"),
63:("ENOSR", "Out of streams resources"),
64:("ENONET", "Machine is not on the network"),
65:("ENOPKG", "Package not installed"),
66:("EREMOTE", "Object is remote"),
67:("ENOLINK", "Link has been severed"),
68:("EADV", "Advertise error"),
69:("ESRMNT", "Srmount error"),
70:("ECOMM", "Communication error on send"),
71:("EPROTO", "Protocol error"),
72:("EMULTIHOP", "Multihop attempted"),
73:("EDOTDOT", "RFS specific error"),
74:("EBADMSG", "Not a data message"),
75:("EOVERFLOW", "Value too large for defined data type"),
76:("ENOTUNIQ", "Name not unique on network"),
77:("EBADFD", "File descriptor in bad state"),
78:("EREMCHG", "Remote address changed"),
79:("ELIBACC", "Can not access a needed shared library"),
80:("ELIBBAD", "Accessing a corrupted shared library"),
81:("ELIBSCN", ".lib section in a.out corrupted"),
82:("ELIBMAX", "Attempting to link in too many shared libraries"),
83:("ELIBEXEC", "Cannot exec a shared library directly"),
84:("EILSEQ", "Illegal byte sequence"),
85:("ERESTART", "Interrupted system call should be restarted"),
86:("ESTRPIPE", "Streams pipe error"),
87:("EUSERS", "Too many users"),
88:("ENOTSOCK", "Socket operation on non-socket"),
89:("EDESTADDRREQ", "Destination address required"),
90:("EMSGSIZE", "Message too long"),
91:("EPROTOTYPE", "Protocol wrong type for socket"),
92:("ENOPROTOOPT", "Protocol not available"),
93:("EPROTONOSUPPORT", "Protocol not supported"),
94:("ESOCKTNOSUPPORT", "Socket type not supported"),
95:("EOPNOTSUPP", "Operation not supported on transport endpoint"),
96:("EPFNOSUPPORT", "Protocol family not supported"),
97:("EAFNOSUPPORT", "Address family not supported by protocol"),
98:("EADDRINUSE", "Address already in use"),
99:("EADDRNOTAVAIL", "Cannot assign requested address"),
100:("ENETDOWN", "Network is down"),
101:("ENETUNREACH", "Network is unreachable"),
102:("ENETRESET", "Network dropped connection because of reset"),
103:("ECONNABORTED", "Software caused connection abort"),
104:("ECONNRESET", "Connection reset by peer"),
105:("ENOBUFS", "No buffer space available"),
106:("EISCONN", "Transport endpoint is already connected"),
107:("ENOTCONN", "Transport endpoint is not connected"),
108:("ESHUTDOWN", "Cannot send after transport endpoint shutdown"),
109:("ETOOMANYREFS", "Too many references: cannot splice"),
110:("ETIMEDOUT", "Connection timed out"),
111:("ECONNREFUSED", "Connection refused"),
112:("EHOSTDOWN", "Host is down"),
113:("EHOSTUNREACH", "No route to host"),
114:("EALREADY", "Operation already in progress"),
115:("EINPROGRESS", "Operation now in progress"),
116:("ESTALE", "Stale NFS file handle"),
117:("EUCLEAN", "Structure needs cleaning"),
118:("ENOTNAM", "Not a XENIX named type file"),
119:("ENAVAIL", "No XENIX semaphores available"),
120:("EISNAM", "Is a named type file"),
121:("EREMOTEIO", "Remote I/O error"),
122:("EDQUOT", "Quota exceeded"),

123:("ENOMEDIUM", "No medium found"),
124:("EMEDIUMTYPE", "Wrong medium type"),
125:("ECANCELED", "Operation Canceled"),
126:("ENOKEY", "Required key not available"),
127:("EKEYEXPIRED", "Key has expired"),
128:("EKEYREVOKED", "Key has been revoked"),
129:("EKEYREJECTED", "Key was rejected by service"),

130:("EOWNERDEAD", "Owner died"),
131:("ENOTRECOVERABLE", "State not recoverable"),

132:("ERFKILL", "Operation not possible due to RF-kill"),
}


ERRNAME = dict(((y[0], (x, y[1])) for x, y in ERRNO.iteritems()))
#EWOULDBLOCK	EAGAIN	# Operation would block
num, msg = ERRNAME["EAGAIN"]
ERRNAME["EWOULDBLOCK"] = (num, "Operation would block")

#EDEADLOCK	EDEADLK
ERRNAME["EDEADLOCK"] = ERRNAME["EDEADLK"]

# man do i agree with you right now
ERRNO[None] = name, msg = ("ENOTENOUGHCAFFINE", "More operator caffine intake required")
ERRNAME[name] = (None, msg)


if __name__ == "__main__":
	from sys import argv

	if len(argv) == 1:
		# print all in order
		for num, (name, msg) in ERRNO.iteritems():
			print "{0:3} {1}: {2}".format(num, name, msg)
	else:
		for val in argv[1:]:
			num, name, msg = get_error(val)
			print "{0:3} {1}: {2}".format(num, name, msg)
